internal/manage/devices/download.go

```diff
@@ -0,0 +1,343 @@
+	headerIndex := make(map[string]int)
+	for i, h := range headers {
+		headerIndex[h] = i
+	}
+
+	// 3. start collecting the data and streaming it to the client
+	w.Header().Set("Content-Disposition", "attachment; filename=devices.csv")
+	w.Header().Set("Content-Type", "text/csv; charset=utf-8")
+	if _, err := fmt.Fprintln(w, strings.Join(headers, ",")); err != nil {
+		return
+	}
+
+	for d := range usefulDevices() {
+		row := make([]string, len(headers))
+		captureMDValues(d, headerIndex, row)
+		for _, t := range d.Traits {
+			info, ok := traitInfo[t.Name]
+			if !ok {
+				continue
+			}
+			values, err := info.get(r.Context(), d.Name)
+			if err != nil {
+				http.Error(w, fmt.Sprintf("failed to get trait info: %q %q %v", d.Name, t.Name, err), http.StatusInternalServerError)
+				continue
+			}
+			for h, v := range values {
+				index, ok := headerIndex[h]
+				if !ok {
+					continue
+				}
+				row[index] = v
+			}
+		}
+		for i, v := range row {
+			row[i] = encodeCsvField(v)
+		}
+		if _, err := fmt.Fprintln(w, strings.Join(row, ",")); err != nil {
+			return
+		}
+	}
+}
+
+func captureMDValues(md *traits.Metadata, headerIndex map[string]int, row []string) {
+	_ = protorange.Range(md.ProtoReflect(), func(values protopath.Values) error {
+		p := values.Path
+		leafStep := p.Index(-1)
+		switch leafStep.Kind() {
+		case protopath.FieldAccessStep:
+			fd := leafStep.FieldDescriptor()
+			if fd == nil {
+				return nil
+			}
+			if fd.Cardinality() == protoreflect.Repeated {
+				return nil // no headers for lists
+			}
+			if fd.Kind() == protoreflect.MessageKind {
+				return nil // no headers for messages
+			}
+		case protopath.ListIndexStep:
+			fd := leafStep.FieldDescriptor()
+			if fd == nil {
+				return nil
+			}
+			if fd.Cardinality() == protoreflect.Repeated {
+				return nil // no headers for lists
+			}
+			if fd.Kind() == protoreflect.MessageKind {
+				return nil // no headers for messages
+			}
+		default:
+		}
+
+		header := protoPathToHeader(p)
+		if header == "" {
+			return nil
+		}
+		index, ok := headerIndex[header]
+		if !ok {
+			return nil // skip
+		}
+		row[index] = values.Values[len(values.Values)-1].String()
+		return nil
+	})
+}
+
+type traitInfo struct {
+	headers []string
+	get     func(ctx context.Context, name string) (map[string]string, error)
+}
+
+func protoPathToHeader(p protopath.Path) string {
+	var parts []string
+	for _, step := range p {
+		switch step.Kind() {
+		case protopath.FieldAccessStep:
+			parts = append(parts, string(step.FieldDescriptor().Name()))
+		case protopath.MapIndexStep:
+			parts = append(parts, step.MapIndex().String())
+		case protopath.ListIndexStep:
+			// skip writing the index, {bar: [{foo}]} -> bar.foo instead of bar[0].foo
+		case protopath.RootStep:
+			// skip writing the root
+		case protopath.UnknownAccessStep:
+		case protopath.AnyExpandStep:
+		}
+	}
+	return strings.Join(parts, ".")
+}
+
+func encodeCsvField(s string) string {
```

Comment:
Why not make use of `encoding/csv` package here? 

```go
        ..
	w.Header().Set("Content-Type", "text/csv")
	w.Header().Set("Content-Disposition", "attachment;filename=data.csv")
        ...
        writer := csv.NewWriter(w)
        
        for d := range usefulDevices() {
            ...
            err := writer.Write(row)
            // handle non-nil err
            writer.Flush()
            ...
        }
```

Reply:
Good point, I will change it